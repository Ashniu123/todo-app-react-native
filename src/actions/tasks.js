import realm from 'realm';
import shortid from 'shortid';

import todoSchema from '../db/schema';

export const FETCH_DATA = 'FETCH_DATA';
export const ADD_ITEM = 'ADD_ITEM';
export const EDIT_ITEM = 'EDIT_ITEM';
export const REMOVE_ITEM = 'REMOVE_ITEM';
export const REMOVE_COMPLETED = 'REMOVE_ALL';
export const RESTORE_STATE = 'RESTORE_STATE';

export const fetchData = (data) => ({
  type: FETCH_DATA,
  data
});

export const startFetchData = () => {
  return (dispatch) => {
    realm
      .open({ schema: [todoSchema] })
      .then((realm) => {
        let fetchedData = realm.objects('Todolist'),
          data = [];
        if (fetchedData.length !== 0) {
          for (let item of fetchedData) {
            data.push({
              key: item.key,
              item: item.item,
              completed: item.completed
            });
          }
        }
        console.log(data);
        dispatch(fetchData(data));
        realm.close();
      })
      .catch((err) => {
        console.log(err);
      });
  };
};

export const addItem = (data) => ({
  type: ADD_ITEM,
  data
});

export const startAddItem = (item) => {
  return (dispatch) => {
    return new Promise((resolve, reject) => {
      realm
        .open({ schema: [todoSchema] })
        .then((realm) => {
          try {
            const data = {
              key: shortid.generate(),
              item,
              completed: false
            };
            realm.write(() => {
              realm.create('Todolist', data);
              console.log('created item');
              dispatch(addItem(data));
              resolve();
            });
          } catch (e) {
            reject('Error on create!');
          } finally {
            realm.close();
          }
        })
        .catch((err) => {
          console.log('Error creating!');
        });
    });
  };
};

export const editItem = (key) => ({
  type: EDIT_ITEM,
  key
});

export const startEditItem = (key) => {
  return (dispatch) => {
    realm
      .open({ schema: [todoSchema] })
      .then((realm) => {
        try {
          realm.write(() => {
            realm.create('Todolist', { key, completed: true }, true);
            console.log('updated item');
            dispatch(editItem(key));
          });
        } catch (e) {
          console.log('Error on edit!');
        } finally {
          realm.close();
        }
      })
      .catch((err) => {
        console.log('Error editing!');
      });
  };
};

export const removeItem = (key) => ({
  type: REMOVE_ITEM,
  key
});

export const startRemoveItem = (key) => {
  return (dispatch) => {
    realm
      .open({ schema: [todoSchema] })
      .then((realm) => {
        try {
          realm.write(() => {
            const itemToDelete = realm.objects('Todolist').filtered(`key = "${key}"`);
            realm.delete(itemToDelete);
            console.log('deleted item');
            dispatch(removeItem(key));
          });
        } catch (e) {
          console.log('Error on removing one!');
        } finally {
          realm.close();
        }
      })
      .catch((err) => {
        console.log('Error Removing!');
      });
  };
};

export const removeCompleted = () => ({
  type: REMOVE_COMPLETED
});

export const startRemoveCompleted = () => {
  return (dispatch) => {
    realm
      .open({ schema: [todoSchema] })
      .then((realm) => {
        try {
          realm.write(() => {
            const allCompletedItems = realm.objects('Todolist').filtered(`completed = true`);
            realm.delete(allCompletedItems);
            console.log('deleted all completed');
            dispatch(removeCompleted());
          });
        } catch (e) {
          console.log('Error on removing all!');
        } finally {
          realm.close();
        }
      })
      .catch((err) => {
        console.log('Error Removing all!');
      });
  };
};

export const restoreState = () => ({
  type: RESTORE_STATE
});

export const startRestoreState = () => {
  return (dispatch, getState) => {
    realm
      .open({ schema: [todoSchema] })
      .then((realm) => {
        try {
          const data = getState().tasks.prev;
          realm.write(() => {
            console.log('pastdata: ', data);
            realm.create('Todolist', data);
            console.log('restored item');
            dispatch(restoreState());
          });
        } catch (e) {
          console.log('Error on restoring state!');
        } finally {
          realm.close();
        }
      })
      .catch((err) => {
        console.log('Error Restoring state');
      });
  };
};
