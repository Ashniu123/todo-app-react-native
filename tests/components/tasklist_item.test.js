import React from 'react';
import { shallow } from 'enzyme';

import TasklistItem from '../../src/components/taskslist/tasklist_item/tasklist_item.component';
import tasks from '../fixtures/tasks';

test('should render TasklistItem component correctly', () => {
	const firstTask = tasks[0];
	const onCheckboxChange = jest.fn();
	const wrapper = shallow(<TasklistItem 
		id={firstTask.key}
		item={firstTask.item}
		completed={firstTask.completed}
		onCheckboxChange={onCheckboxChange}
	/>);

	expect(wrapper).toMatchSnapshot();
});