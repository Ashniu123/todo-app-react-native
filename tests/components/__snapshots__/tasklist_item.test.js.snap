// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render TasklistItem component correctly 1`] = `
<Component
  style={
    Object {
      "backgroundColor": "#c62828",
      "justifyContent": "center",
      "marginLeft": -100,
    }
  }
>
  <AnimatedComponent
    onMoveShouldSetResponder={[Function]}
    onMoveShouldSetResponderCapture={[Function]}
    onResponderEnd={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderReject={[Function]}
    onResponderRelease={[Function]}
    onResponderStart={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    onStartShouldSetResponderCapture={[Function]}
    style={
      Array [
        Object {
          "left": 0,
          "top": 0,
        },
      ]
    }
  >
    <Component
      style={
        Object {
          "alignItems": "center",
          "bottom": 0,
          "flexDirection": "row",
          "justifyContent": "flex-end",
          "left": 0,
          "position": "absolute",
          "top": 0,
          "width": 100,
        }
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#ffffff",
            "margin": 16,
          }
        }
      >
        DELETE
      </Text>
    </Component>
    <Component
      style={
        Array [
          Object {
            "flexDirection": "row",
            "justifyContent": "space-between",
            "marginLeft": 100,
            "padding": 10,
          },
          Object {
            "backgroundColor": "#ffffff",
          },
        ]
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#212121",
            "fontSize": 15,
          }
        }
      >
        Complete writing all test cases
      </Text>
      <TouchableWithoutFeedback
        hitSlop={
          Object {
            "bottom": 10,
            "left": 10,
            "right": 10,
            "top": 10,
          }
        }
      >
        <Icon
          color="black"
          name="checkbox-blank-circle-outline"
          onPress={[Function]}
          raised={false}
          reverse={false}
          reverseColor="white"
          size={24}
          type="material-community"
          underlayColor="white"
        />
      </TouchableWithoutFeedback>
    </Component>
  </AnimatedComponent>
</Component>
`;
